src/coreclr/vm/zapsig.cpp

```diff
@@ -343,7 +340,7 @@ BOOL ZapSig::GetSignatureForTypeHandle(TypeHandle      handle,
         {
             DWORD ix = CorSigUncompressData(pSig);
             CONTRACT_VIOLATION(ThrowsViolation|GCViolation);
-            pModule = pZapSigContext->GetZapSigModule()->GetModuleFromIndexIfLoaded(ix);
+            pModule = pZapSigContext->GetZapSigModule()->GetModuleFromIndex(ix);
```

Comment:
Can this cause random assemblies to get loaded early?

Reply:
Well, with a probability corresponding to hash collisions (theoretically 2^-32) it can load more assemblies than required. One possible way to mitigate this would be to add a third method, something like <code>GetModuleFromIndexIfLoadedOrComposite</code>, that would only apply the logic of loading extra assemblies in the presence of a composite image thus removing any impact on non-composite code (even though in practice I doubt this happens very often as single-assembly crossgen files shouldn't precompile generic instantiations involving references to other modules. Without some form of loading the reference assembly I currently don't see how to make the precompiled generic instantiations in composite images useful.

Reply:
> Well, with a probability corresponding to hash collisions (theoretically 2^-32)

I believe that the probability can be much higher in aggregate due birthday paradox.

>  Without some form of loading the reference assembly I currently don't see how to make the precompiled generic instantiations in composite images useful.

I think you just need to resolve the reference assembly from the already loaded assemblies to see whether it matches. All constituents of the generic method that we are trying to look up are loaded already. It means that we are not going to miss anything when we do not load any new assemblies.

Reply:
Fixed in 2nd commit, thanks for pointing that out.